### PR TITLE
Modify logic for binary operations on Slate tensors

### PR DESCRIPTION
**Pull request by thomasgibson**
I ran into this problem while working on some HDG stuff. There are instances when using `tensor` blocking and multiplying against an `AssembledVector`, for example, causes a `ValueError` because the contracted spaces are not the "same."

This occurs because calling `block` on a tensor reconstructs function spaces such that they agree with the argument indices of the underlying UFL form, or other Slate objects. (Similar to what is done when you call the form splitter). Blocking does not carry over the information that a particular function space is an indexed subspace of a mixed space. Therefore, I run into instances where I hit my own error-checking when performing a Slate mat-vec, even though it mathematically makes sense.

This reproduces the error:
```python
from firedrake import *


mesh = UnitIcosahedralSphereMesh(3)

V = VectorFunctionSpace(mesh, "DG", 1)
U = FunctionSpace(mesh, "DG", 1)
T = FunctionSpace(mesh, "DGT", 1)
W = V * U * T

w, psi, gammar = TestFunctions(W)
u, phi, lambdar = TrialFunctions(W)

a = inner(w, u)*dx + inner(psi, phi)*dx + gammar('+')*lambdar('+')*dS

A = Tensor(a)

w = Function(W)
_, _, lambda_hyb = w.split()

A02 = A.block((0, 2))

# Slate coefficient vector for lambda.
# Note: same function space as the trial space for A02 (but is indexed)
Lambda = AssembledVector(lambda_hyb)
foo = A02 * Lambda
```

I'm not sure how to correct this without modifying the current logic for binary operations on Slate expressions. It wasn't invasive and this change fixes the issue. I'm open to suggestions.

### CODE REVIEW

File: firedrake/slate/slate.py [hunk 1]
```diff
@@ -821,6 +825,24 @@ def arguments(self):
...
+    :arg A: A function space.
+    :arg B: Another function space.
+
+    Returns `True` if they have matching meshes, elements, and rank. Otherwise,
+    `False` is returned.
+    """
+
+    equiv = False
+    assert A.mesh() == B.mesh(), "Mismatching meshes!"
+    if A.rank == B.rank and A.ufl_element() == B.ufl_element():
```
Comment (wence-) [inline firedrake/slate/slate.py]: Under what circumstances do the elements match, but the rank not?

File: firedrake/slate/slate.py [hunk 2]
```diff
@@ -821,6 +825,24 @@ def arguments(self):
...
+
+    :arg A: A function space.
+    :arg B: Another function space.
+
+    Returns `True` if they have matching meshes, elements, and rank. Otherwise,
+    `False` is returned.
+    """
+
+    equiv = False
+    assert A.mesh() == B.mesh(), "Mismatching meshes!"
```
Comment (wence-) [inline firedrake/slate/slate.py]: I would allow checking on mismatching meshes, but say false.

File: firedrake/slate/slate.py [hunk 3]
```diff
@@ -821,6 +825,24 @@ def arguments(self):
...
+    if A.rank == B.rank and A.ufl_element() == B.ufl_element():
+        equiv = True
+
+    return equiv
```
Comment (wence-) [inline firedrake/slate/slate.py]: You can do beta-reduction:

`return A.mesh() == B.mesh() and A.ufl_element() == B.ufl_element() and A.rank == B.rank`.

File: firedrake/slate/slate.py (hunk 1 above)
Comment (thomasgibson) [reply firedrake/slate/slate.py]: This is just me being paranoid and making *extra* sure. But now that I think about it, rank is probably a superfluous check.

**Comment by thomasgibson** [comment]
Good to go?

**Comment by wence-** [comment]
OK.

**Pull request merged**